File: dummy_symfony/src/Service/ISSService.php
<?php
namespace App\Service;

use Exception;

class ISSService
{
    private $issApiUrl = 'http://api.open-notify.org/iss-now.json';
    private $nominatimUrl = 'https://nominatim.openstreetmap.org/reverse';
    private $userAgent = 'iss-symfony-app';

    public function getIssPosition(): array
    {
        $response = file_get_contents($this->issApiUrl);
        if ($response === false) {
            throw new Exception('Failed to fetch ISS position');
        }
        $data = json_decode($response, true);
        $lat = $data['iss_position']['latitude'] ?? null;
        $lon = $data['iss_position']['longitude'] ?? null;
        if (!$lat || !$lon) {
            throw new Exception('Invalid ISS position data');
        }
        return [$lat, $lon];
    }

    public function getLocationInfo(string $lat, string $lon): array
    {
        $params = http_build_query([
            'format' => 'json',
            'lat' => $lat,
            'lon' => $lon,
            'zoom' => 10,
            'addressdetails' => 1
        ]);
        $opts = [
            'http' => [
                'header' => "User-Agent: {$this->userAgent}\r\n"
            ]
        ];
        $context = stream_context_create($opts);
        $url = $this->nominatimUrl . '?' . $params;
        $response = file_get_contents($url, false, $context);
        if ($response === false) {
            return ['type' => 'unknown', 'name' => 'Unknown'];
        }
        $data = json_decode($response, true);
        $address = $data['address'] ?? [];
        if (isset($address['country'])) {
            return ['type' => 'country', 'name' => $address['country']];
        }
        foreach (['sea', 'ocean', 'water', 'river', 'lake'] as $key) {
            if (isset($address[$key])) {
                return ['type' => 'sea', 'name' => $address[$key]];
            }
        }
        return ['type' => 'unknown', 'name' => 'Unknown'];
    }

    public function getIssPositionWithLocation(): array
    {
        list($lat, $lon) = $this->getIssPosition();
        $location = $this->getLocationInfo($lat, $lon);
        return [
            'latitude' => $lat,
            'longitude' => $lon,
            'location_type' => $location['type'],
            'location_name' => $location['name']
        ];
    }
} 

File: dummy_microservices_v2/backend/app/iss_service.py
import requests
import logging
from typing import Dict

logger = logging.getLogger("iss_service")

class ISSService:
    """
    Service for fetching ISS position and resolving country or sea/ocean.
    """
    ISS_API_URL = "http://api.open-notify.org/iss-now.json"
    NOMINATIM_URL = "https://nominatim.openstreetmap.org/reverse"
    USER_AGENT = {"User-Agent": "iss-fastapi-app"}

    def get_iss_position(self) -> Dict[str, str]:
        """
        Fetch the current ISS position (latitude, longitude).
        Logs the API call and result.
        """
        logger.info("Fetching ISS position from %s", self.ISS_API_URL)
        response = requests.get(self.ISS_API_URL, timeout=5)
        response.raise_for_status()
        iss_data = response.json()
        lat = iss_data["iss_position"]["latitude"]
        lon = iss_data["iss_position"]["longitude"]
        logger.info("ISS position: latitude=%s, longitude=%s", lat, lon)
        return {"latitude": lat, "longitude": lon}

    def get_location_info(self, lat: str, lon: str) -> Dict[str, str]:
        """
        Given latitude and longitude, return the country or sea/ocean name.
        Logs the API call, result, and errors.
        """
        params = {
            "format": "json",
            "lat": lat,
            "lon": lon,
            "zoom": 10,
            "addressdetails": 1
        }
        try:
            logger.info("Resolving location for lat=%s, lon=%s", lat, lon)
            response = requests.get(self.NOMINATIM_URL, headers=self.USER_AGENT, params=params, timeout=5)
            response.raise_for_status()
            data = response.json()
            address = data.get("address", {})
            if "country" in address:
                logger.info("Location resolved: country=%s", address["country"])
                return {"type": "country", "name": address["country"]}
            for key in ["sea", "ocean", "water", "river", "lake"]:
                if key in address:
                    logger.info("Location resolved: sea=%s", address[key])
                    return {"type": "sea", "name": address[key]}
            logger.warning("Location unknown for lat=%s, lon=%s", lat, lon)
            return {"type": "unknown", "name": "Unknown"}
        except Exception as e:
            logger.error("Error resolving location for lat=%s, lon=%s: %s", lat, lon, e)
            return {"type": "unknown", "name": "Unknown"}

    def get_iss_position_with_location(self) -> Dict[str, str]:
        """
        Fetch the ISS position and resolve the country or sea/ocean.
        Logs the combined result.
        """
        pos = self.get_iss_position()
        location = self.get_location_info(pos["latitude"], pos["longitude"])
        result = {**pos, "location_type": location["type"], "location_name": location["name"]}
        logger.info("ISS position with location: %s", result)
        return result 